### PR TITLE
Additional filter functions

### PR DESCRIPTION
**Pull request by florian-huber**
@kiaashour added 4 new filters (see #107 ):

- ``remove_peaks_around_precursor_mz``
- ``remove_peaks_outside_top_k``
- ``require_minimum_of_high_peaks``
- ``require_precursor_below_mz``

### CODE REVIEW

**Comment by florian-huber** [comment]
Hi @kiaashour , I again renamed 2 of the functions to make them consistent with ``require_minimum_number_of_peaks`` filter.
For the tests you could have a look at ``filter\test_require_minimum_number_of_peaks.py`` which should be easy to adapt for some of the new functions.

**Comment by kiaashour** [comment]
Great. I've began writing the tests and should be able to upload them soon.

**Comment by kiaashour** [comment]
Hi Florian. I've written some tests (one still remaining) and updated the filtering functions a bit, and was just wondering how I should push them to the PR. Right now I've sent another PR but I believe that's not needed?

**Comment by florian-huber** [comment]
Thanks Kia for adding tests!
I gave you writing rights, so we should now be able to both work on this PR (and I will merge your changes into this one).

File: matchms/filtering/require_precursor_below_mz.py
Comment (florian-huber) [inline matchms/filtering/require_precursor_below_mz.py]:
```suggestion
    if precursor_mz and precursor_mz >= max_mz:
        return None
```

Comment (florian-huber) [reply matchms/filtering/require_precursor_below_mz.py]: SonarCloud complains about a "code smell" which is the tools' way of suggesting code improvements. I think the suggested change would solve that.

**Comment by florian-huber** [comment]
Looking good! Just push the missing test and then SonarCloud should be happy (wants code coverage >= 80%).

By the way, SonarCloud is worth checking out! If you click on ``Coverage`` or ``Code smell`` it usually does a good job at pointing to thinks that can be improved.

We should also update the Changelog.

**Comment by kiaashour** [comment]
Great, I'll add the remaining test now and have a look at the Changelog. Will keep an eye on SonarCloud too.

File: matchms/filtering/require_precursor_below_mz.py
Comment (kiaashour) [reply matchms/filtering/require_precursor_below_mz.py]: My reason for the nested if statements was in case the precursor_mz was missing (so would be None, rather than 0).  But I suppose we can ignore that?

Comment (florian-huber) [reply matchms/filtering/require_precursor_below_mz.py]: That will be dealt with by Python. 
If precursor_mz is None than it will not go further to check the ``and ...`` part.
(so if you change the order: ``if precursor_mz >= max_mz and precursor_mz`` that would indeed fail for ``None``)

Comment (kiaashour) [reply matchms/filtering/require_precursor_below_mz.py]: Ah I see, so it wouldn't give a TypeError.

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (florian-huber) [inline matchms/filtering/remove_peaks_around_precursor_mz.py]:
```suggestion
    if precursor_mz:
        peaks_to_remove = ((numpy.abs(precursor_mz-mzs) <= mz_tolerance) & (mzs != precursor_mz))
        new_mzs, new_intensities = mzs[~peaks_to_remove], intensities[~peaks_to_remove]
```

Comment (florian-huber) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: I guess that a numpy way of implementing this might be faster because it won't need the loop. And makes it a bit simpler code-wise. Should be doing exactly the same thing (I hope).

File: matchms/filtering/remove_peaks_outside_top_k.py
Comment (florian-huber) [inline matchms/filtering/remove_peaks_outside_top_k.py]:
```suggestion
    keep_idx = [i for i in top_k]
    for i in indices:

        compare = abs(mzs[i]-k_ordered_mzs) <= mz_window
        if numpy.any(compare):
            keep_idx.append(i)

    keep_idx.sort()
    new_mzs, new_intensities = mzs[keep_idx], intensities[keep_idx]
```

Comment (florian-huber) [reply matchms/filtering/remove_peaks_outside_top_k.py]: I would rather avoid using the NaN's here as markers.

**Comment by florian-huber** [comment]
Hi @kiaashour . Thanks for adding the missing test!
I went over the functions again and made some suggestions to avoid using NaN's as markers.

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (kiaashour) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Yeah works just the same. Thanks.

**Comment by florian-huber** [comment]
Thanks for the changes Kia. I updated the changelog and merged master into this branch.

**Comment by kiaashour** [comment]
Great, thank you. Shall we wait for @sverhoeven to review before merging?

**Comment by florian-huber** [comment]
Yes, we should wait for the review. That's usually a very good last check before merging.

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (sverhoeven) [inline matchms/filtering/remove_peaks_around_precursor_mz.py]: If there is no precursor then spectrum is returned. I expected an exception to be thrown because the filter requires a precursor. See for example [add_losses()](https://github.com/matchms/matchms/blob/master/matchms/filtering/add_losses.py#L25)

File: tests/test_remove_peaks_around_precursor_mz.py
Comment (sverhoeven) [inline tests/test_remove_peaks_around_precursor_mz.py]: Use defaults instead of no parameters.
```suggestion
def test_remove_peaks_around_precursor_mz_defaults_no_precursor():
     """Using defaults and no precursor mz present."""
```

File: matchms/filtering/remove_peaks_outside_top_k.py
Comment (sverhoeven) [inline matchms/filtering/remove_peaks_outside_top_k.py]: Dont understand what the mz window does. 
1. If one top peak has a peak (outside top k) near it within the mz window will it be kept or removed?
2. If 2 top k peaks are within the window will one be discarded?

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Please clone Spectrum instead of overwriting in-place in all the functions. In all other filtering function we use a cloned spectrum a we want to treat the Spectrum objects as immutable.
```python
spectrum = spectrum_in.clone()
```

**Comment by florian-huber** [comment]
> 
> 
> Please clone Spectrum instead of overwriting in-place in all the functions. In all other filtering function we use a cloned spectrum a we want to treat the Spectrum objects as immutable.
> 
> ```python
> spectrum = spectrum_in.clone()
> ```

Good catch!
I added a test to all four test files to check if the input spectrum is properly cloned.

File: matchms/filtering/remove_peaks_outside_top_k.py
Comment (florian-huber) [reply matchms/filtering/remove_peaks_outside_top_k.py]: @kiaashour can you edit the description in the docstring?

To Stefans questions: 1. that will be kept. 2. no, all top k peaks are kept.

My docstring suggestion: ``"""Remove all peaks which are not within *mz_window* of at least one of the *k* highest intensity peaks of the spectrum.``

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (florian-huber) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Just saw that also in ``add_losses()`` we have the assert after ``if precursor_mz``. 
We can 1) add the same assert function here. In that case we accept missing precursor_mz entries, but if there is an entry it must be a number.
What you suggest here is 2) add the assert before the if and make the function fail when there is no numerical ``precursor_mz`` (right?). To me that's fine as well and indeed seems cleaner. We just would have to check if that breaks something. And if we do, we should do it for both this function and ``add_losses``.

Comment (kiaashour) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: 2) would be clearer to me also. Can't really see it breaking anything in the filter function or add_losses.

**Comment by kiaashour** [comment]
I think I will also add a check for when the input spectra are None as well as that's been done on all the other filter functions.

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (florian-huber) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Let's do 2 then.

Comment (kiaashour) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: I think we should also add the same thing for `require_precursor_below_mz` for consistency.

Comment (florian-huber) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Definitely! Good point.

Comment (kiaashour) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Some of the tests fail when we add an assertion for the missing precursor mz. For `require_precursor_below_mz` the cloning test fails (raises an assertion error for missing precursor mz) and the test for when the precursor mz is absent fails as well. I suppose we can just remove the test for missing precursor mz and add a dummy precursor mz to the cloning test?

Comment (florian-huber) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Yes, we can simply add a precursor_mz to the ``test_require_precursor_below_mz``.

Comment (kiaashour) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Shall I also make the same changes for `add_losses`?

Comment (florian-huber) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: I'll change add_losses in a separate, tiny PR, just to not make this PR even bigger. (I also added an issue for that).

**Comment by florian-huber** [comment]
Thanks for the changes @kiaashour !
I just added a few tests to check if the assert messages are raised correctly and asked Stefan to check again if his comments were all addressed.

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (sverhoeven) [inline matchms/filtering/remove_peaks_around_precursor_mz.py]: Match docstring with function spec
```suggestion
    spectrum_in
```

File: matchms/filtering/remove_peaks_outside_top_k.py
Comment (sverhoeven) [inline matchms/filtering/remove_peaks_outside_top_k.py]: Match docstring with function spec
```suggestion
    spectrum_in
```

File: matchms/filtering/require_minimum_of_high_peaks.py
Comment (sverhoeven) [inline matchms/filtering/require_minimum_of_high_peaks.py]: Match docstring with function spec
```suggestion
    spectrum_in
```

File: matchms/filtering/require_precursor_below_mz.py
Comment (sverhoeven) [inline matchms/filtering/require_precursor_below_mz.py]: Match docstring with function spec
```suggestion
    spectrum_in
```

**Review by sverhoeven — APPROVED** [review]
Nice work. 100% coverage and useful additions to library.

Just some small docstring fixes.

File: matchms/filtering/remove_peaks_around_precursor_mz.py
Comment (kiaashour) [reply matchms/filtering/remove_peaks_around_precursor_mz.py]: Ok great.

**Comment by kiaashour** [comment]
Shall we just keep both changes from master and this branch for the conflict, seeing that we'll be merging this into master?

**Comment by florian-huber** [comment]
@kiaashour I indeed kept both changes in the changelog. 

I also added you to the list of contributors (``.zenodo.json`` and ``CITATION.cff``). Please check if your name and affiliation is correct, and add an orcid if you have one. Once that is done I'll merge it into master.

**Comment by kiaashour** [comment]
Great, thank you. Looks all fine to me, added my orcid ID as well.

**Comment by florian-huber** [comment]
Thanks a lot @kiaashour and @sverhoeven , great work!

**Pull request merged**
